### DISCUSSION TITLE
Why is Ocelot ignoring the first route in my JSON file?

### DISCUSSION THREAD

**Discussion by dcmeadley** [ThreeMammals/Ocelot#1855] [Q&A] [answered]
I have implemented Ocelot API Gateway in my Net Core 7 application.  One of the JSON files containing routes is set up as follows:

```json
{
  "Routes": [
    {
      "DownstreamPathTemplate": "/storedfiletype",
      "DownstreamScheme": "https",
      "DownstreamHostAndPorts": [
        {
          "Host": "<<<HOST DELETED FOR SECURITY>>>",
          "Port": 443
        }
      ],
      "UpstreamPathTemplate": "/fileserver/storedfiletype",
      "UpstreamHttpMethod": [ "Post" ]
    },
    {
      "DownstreamPathTemplate": "/storedfiletype/fileextension",
      "DownstreamScheme": "https",
      "DownstreamHostAndPorts": [
        {
          "Host": "<<<HOST DELETED FOR SECURITY>>>",
          "Port": 443
        }
      ],
      "UpstreamPathTemplate": "/fileserver/storedfiletype/fileextension",
      "UpstreamHttpMethod": [ "Post" ]
    }
  ]
}
```

Now when I try and post to the first route (```/storedfiletype``` in this case), I get a 404 error but the second route works.  When I swap them over (making ```/storedfiletype/fileextension``` the first route), ```/storedfiletype``` works and ```/storedfiletype/fileextension``` gives me the 404 error.

Here are the logs I am getting:
```
Request starting HTTP/1.1 POST https://<<<HOST DELETED FOR SECURITY>>>/fileserver/storedfiletype application/json 199
Wildcard detected, all requests with hosts will be allowed.
requestId: 80001b3a-100c-8000-b63f-84710c7967bb, previousRequestId: no previous request id, message: ocelot pipeline started
requestId: 80001b3a-100c-8000-b63f-84710c7967bb, previousRequestId: no previous request id, message: Upstream url path is /fileserver/storedfiletype
requestId: 80001b3a-100c-8000-b63f-84710c7967bb, previousRequestId: no previous request id, message: DownstreamRouteFinderMiddleware setting pipeline errors. IDownstreamRouteFinder returned Error Code: UnableToFindDownstreamRouteError Message: Failed to match Route configuration for upstream path: /fileserver/storedfiletype, verb: POST.
requestId: 80001b3a-100c-8000-b63f-84710c7967bb, previousRequestId: no previous request id, message: Error Code: UnableToFindDownstreamRouteError Message: Failed to match Route configuration for upstream path: /fileserver/storedfiletype, verb: POST. errors found in ResponderMiddleware. Setting error response for request path:/fileserver/storedfiletype, request method: POST
requestId: 80001b3a-100c-8000-b63f-84710c7967bb, previousRequestId: no previous request id, message: ocelot pipeline finished
Request finished HTTP/1.1 POST https://api.mnet.network/fileserver/storedfiletype application/json 199 - 404 - - 69.5534ms
```

**Comment by raman-m** [comment, marked as answer]
Hi @dcmeadley !
Thanks for your interest in Ocelot gateway!

I love your text replacement placeholder `<<<HOST DELETED FOR SECURITY>>>` :rofl: 

After review of your routes and exception which is:
```
requestId: 80001b3a-100c-8000-b63f-84710c7967bb,
previousRequestId: no previous request id, 
message: Error Code: UnableToFindDownstreamRouteError 
Message: Failed to match Route configuration for upstream path: /fileserver/storedfiletype, verb: POST. errors found in ResponderMiddleware. 
Setting error response for request path:/fileserver/storedfiletype, request method: POST
```
...I would say that root cause is `Failed to match Route configuration for upstream path: /fileserver/storedfiletype`.
This means that you've defined two routes with **the same template for the same downservice endpoint with the same base path**! You've violated configuration requirements for [Routing](https://ocelot.readthedocs.io/en/latest/features/routing.html) feature. So, **the second route is duplicate!**

For example, if the route prefix is the same for some routes then you have to understand [Catch All](https://ocelot.readthedocs.io/en/latest/features/routing.html#catch-all) approach, and merge all related routes to one CatchAll-route. 

To fix the error you have to merge both routes, and introduce one `{everything}`-route like that:
```json
{
  "Routes": [
    {
      "DownstreamPathTemplate": "/storedfiletype/{everything}",
      "DownstreamScheme": "https",
      "DownstreamHostAndPorts": [
        {
          "Host": "<<<HOST DELETED FOR SECURITY>>>",
          "Port": 443
        }
      ],
      "UpstreamPathTemplate": "/fileserver/storedfiletype/{everything}",
      "UpstreamHttpMethod": [ "Post" ]
    }
  ]
}
```
Or even shorter template:
```json
  "UpstreamPathTemplate": "/fileserver/{everything}",
```
Hope you will provide a feedback soon!

**Comment by raman-m** [comment]
> I have implemented Ocelot API Gateway in my **Net Core 7** application. 

We have the latest .NET 7 [release](../releases) with [19.0.2](https://www.nuget.org/packages/Ocelot) version, so you can upgrade your gateway solution. :wink:

**Reply by raman-m** [reply]
Dear @dcmeadley !
### Update # 1
We have latest [22.0.1](https://github.com/ThreeMammals/Ocelot/releases) version to upgrade to!
